File: 2025/luke_4/zig/main.zig
const std = @import("std");

fn countNeighbors(map: *std.AutoHashMap(usize, []const u8), row: usize, col: usize, target: u8) usize {
    const directions = [_][2]i32{
        .{ -1, -1 }, .{ -1, 0 }, . { -1, 1 }, // top-left, top, top-right
        .{ 0, - 1 },             . { 0, 1  }, // left, right
        .{ 1, - 1 }, .{ 1, 0  },  .{ 1, 1  }, // bottom-left, bottom, bottom-right
    };

    var count: usize = 0;

    for (directions) |dir| {
        const new_row = @as(i32, @intCast(row)) + dir[0];
        const new_col = @as(i32, @intCast(col)) + dir[1];

        if (new_row < 0 or new_col < 0) continue;

        const r = @as(usize, @intCast(new_row));
        const c = @as(usize, @intCast(new_col));

        if (map.get(r)) |line| {
            if (c < line.len and line[c] == target) {
                count += 1;
            }
        }
    }

    return count;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    var map = std.AutoHashMap(usize, []const u8).init(allocator);
    defer map.deinit();

    var read_buf: [4096]u8 = undefined;
    var file_reader = file.reader(&read_buf);

    var row: usize = 0;
    while (try file_reader.interface.takeDelimiter('\n')) |line| {
        const line_copy = try allocator.dupe(u8, line);
        try map.put(row, line_copy);
        row += 1;
    }

    var task1: usize = 0;
    var it = map.iterator();
    while (it.next()) |entry| {
        const row_idx = entry.key_ptr.*;
        const line = entry.value_ptr.*;

        for (line, 0..) |char, col_idx| {
            if (char != '@') {
                continue;
            }
            const count = countNeighbors(&map, row_idx, col_idx, '@');
            if (count < 4) {
                task1 += 1;
            }
        }
    }

    var task2: usize = 0;
    var changed: bool = true;
    while (changed) {
        var newit = map.iterator();
        changed = false;
        while (newit.next()) |entry| {
            const row_idx = entry.key_ptr.*;
            const line = entry.value_ptr.*;

            for (line, 0..) |char, col_idx| {
                if (char != '@') {
                    continue;
                }
                const count = countNeighbors(&map, row_idx, col_idx, '@');
                if (count < 4) {
                    task2 += 1;
                    const mutable_line = @constCast(line);
                    mutable_line[col_idx] = '.';
                    changed = true;
                }
            }
        }
    }

    std.debug.print("task1: {}\n", .{task1});
    std.debug.print("task2: {}\n", .{task2});

    var it2 = map.valueIterator();
    while (it2.next()) |value| {
        allocator.free(value.*);
    }
}

File: 2025/luke_4/rust/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open("../input.txt")?;
    let reader = BufReader::new(file);

    let mut factory: Vec<Vec<char>> = Vec::with_capacity(140);
    for line in reader.lines() {
        let line = line?;
        factory.push(line.chars().collect());
    }

    let mut task1 = 0;
    for y in 0..factory.len() {
        for x in 0..factory[y].len() {
            if factory[y][x] != '@' {
                continue;
            }
            let num = adjacent_rolls(&factory, x, y);
            if num < 4 {
                task1 += 1;
            }
        }
    }

    let mut task2 = 0;
    let mut changed = true;
    while changed {
        changed = false;
        for y in 0..factory.len() {
            for x in 0..factory[y].len() {
                if factory[y][x] != '@' {
                    continue;
                }
                let num = adjacent_rolls(&factory, x, y);
                if num < 4 {
                    task2 += 1;
                    factory[y][x] = '.';
                    changed = true;
                }
            }
        }
    }

    println!("task1 {}", task1);
    println!("task2 {}", task2);

    Ok(())
}

fn adjacent_rolls(factory: &[Vec<char>], x: usize, y: usize) -> i32 {
    let mut count = 0;
    let max_y = factory.len();

    let directions = [
        (-1, -1),
        (0, -1),
        (1, -1),
        (-1, 0),
        (1, 0),
        (-1, 1),
        (0, 1),
        (1, 1),
    ];

    for (dx, dy) in directions {
        let new_y = y as i32 + dy;
        let new_x = x as i32 + dx;

        if new_y < 0 || new_y >= max_y as i32 || new_x < 0 {
            continue;
        }

        let row = &factory[new_y as usize];
        if new_x < row.len() as i32 && row[new_x as usize] == '@' {
            count += 1;
        }
    }

    count
}
